pennylane/shadows/classical_shadow.py

```diff
@@ -404,42 +403,54 @@ def qnode(x):
             bitstrings, recipes = qnode(x)
             shadow = qml.ClassicalShadow(bitstrings, recipes)
 
-        >>> [shadow.entropy(wires=wires, alpha=alpha, atol=1e-10) for alpha in [1., 2., 3.]]
+        >>> [shadow.entropy(wires=wires, alpha=alpha) for alpha in [1., 2., 3.]]
         [1.5419292874423107, 1.1537924276625828, 0.9593638767763727]
 
         """
 
+        if atol is not None:
+            warnings.warn(
+                "Trying to use atol in ClassicalShadow.entropy. This feature is deprecated and has been replaced by a more accurate version that projects the density matrix estimator to the closest valid density matrix in L2 norm. You can simply remove the atol keyword in your code.",
+                qml.PennyLaneDeprecationWarning,
+            )
+
         global_snapshots = self.global_snapshots(wires=wires, snapshots=snapshots)
         rdm = median_of_means(global_snapshots, k, axis=0)
 
         # Allow for different log base
         div = np.log(base) if base else 1
 
-        # This was returning negative values in most cases, so commenting it out
-        # until we figure out the issue.
-        # if alpha == 2:
-        #     # special case of purity
-        #     res = -qml.math.log(qml.math.trace(rdm @ rdm))
-        #     return res / div
-
-        # Else
-        # Compute Eigenvalues and choose only those >>0
-        evs = qml.math.eigvalsh(rdm)
-        mask0 = qml.math.logical_not(qml.math.isclose(evs, 0, atol=atol))
-        mask1 = qml.math.where(evs > 0, True, False)
-        mask = qml.math.logical_and(mask0, mask1)
-        # Renormalize because of cropped evs
-        evs_nonzero = qml.math.gather(evs, mask)
-        evs_nonzero = evs_nonzero / qml.math.sum(evs_nonzero)
-
+        evs_nonzero = _project_density_matrix_spectrum(rdm)
         if alpha == 1:
             # Special case of von Neumann entropy
-            return -qml.math.sum(evs_nonzero * qml.math.log(evs_nonzero)) / div
+            return qml.math.entr(evs_nonzero) / div
 
         # General Renyi-alpha entropy
         return qml.math.log(qml.math.sum(evs_nonzero**alpha)) / (1.0 - alpha) / div
 
 
+def _project_density_matrix_spectrum(rdm):
+    """Project the estimator density matrix rdm with possibly negative eigenvalues onto the closest true density matrix in L2 norm"""
+    # algorithm below eq. (16) in https://arxiv.org/pdf/1106.5458.pdf
+    evs = qml.math.eigvalsh(rdm)[::-1]  # order from largest to smallest
+    d = len(rdm)
+    a = 0.0
+    break_happened = False
+    for i in range(d - 1, 0, -1):
+        if evs[i] + a / (i + 1) > 0:
+            break_happened = True
+            break
+        a += evs[i]
+
+    if not break_happened:
+        # Edge case: when no break happened the closest spectrum is trivially just [1.,]
+        return qml.math.ones_like(evs)[:1]
+
+    lambdas = qml.math.zeros_like(evs)[: i + 1]  # only keep non-zero ones
+    lambdas = evs[: i + 1] + a / (i + 1)
+    return lambdas[::-1]
```

Comment:
```suggestion
    evs = qml.math.eigvalsh(rdm)[::-1]  # order from largest to smallest
    d = len(rdm)
    a = 0.0
    for i in range(d - 1, -1, -1):
        if evs[i] + a / (i + 1) > 0:
            break
        a += evs[i]

    lambdas = qml.math.zeros_like(evs)[: i + 1]  # only keep non-zero ones
    lambdas = evs[: i + 1] + a / (i + 1)
    return lambdas[::-1]
```
Okay now I kind of feel like I'm just being difficult, and it seems like this does work, but looking at it again, I think once the index in line 440 was updated to be `i+1`, all of this edge case business was handled automatically by the procedure from the paper, so I don't think this is necessary. I believe the problem before was that the break wasn't happening for the final value because weird things happened with `evs[i] + a / i  > 0` for `i=0`. With the switch to `i+1`, the break happens before the final eigenvalue is added to `a`, and everything works out.

I made the suggested change locally (removed the `break_happened` logic, and changed the end index in range back to -1) and all the tests still pass, including the new trivial case examples.

Reply:
Oh I thought I had checked that but maybe I forgot to set the range back to -1. If this works then great! Thanks a lot!